pageserver/src/tenant_threads.rs

```diff
@@ -1,77 +1,173 @@
+        .thread_name("tenant-task-worker")
+        .worker_threads(40) // Way more than necessary
+        .max_blocking_threads(100) // Way more than necessary
+        .enable_all()
+        .build()?;
+
+    let (gc_send, mut gc_recv) = mpsc::channel::<ZTenantId>(100);
+    START_GC_LOOP.set(gc_send).unwrap();
+
+    let (compaction_send, mut compaction_recv) = mpsc::channel::<ZTenantId>(100);
+    START_COMPACTION_LOOP.set(compaction_send).unwrap();
+
+    thread_mgr::spawn(
+        ThreadKind::TenantTaskManager,
+        None,
+        None,
+        "WAL receiver manager main thread",
+        true,
+        move || {
+            runtime.block_on(async move {
+                loop {
+                    tokio::select! {
+                        _ = thread_mgr::shutdown_watcher() => break,
+                        // TODO don't spawn if already running
+                        tenantid = gc_recv.recv() => {
+                            tokio::spawn(gc_loop(tenantid.unwrap()));
+                        },
+                        tenantid = compaction_recv.recv() => {
+                            tokio::spawn(compaction_loop(tenantid.unwrap()));
+                        },
+                    }
+                }
+            });
+            Ok(())
+        },
+    )?;
+
+    Ok(())
+}
+
 ///
 /// GC thread's main loop
 ///
-pub fn gc_loop(tenantid: ZTenantId) -> Result<()> {
+async fn gc_loop(tenantid: ZTenantId) -> Result<()> {
     loop {
         if tenant_mgr::get_tenant_state(tenantid) != Some(TenantState::Active) {
             break;
         }
 
-        trace!("gc thread for tenant {} waking up", tenantid);
+        trace!("gc loop for tenant {} waking up", tenantid);
         let repo = tenant_mgr::get_repository_for_tenant(tenantid)?;
+        let gc_period = repo.get_gc_period();
         let gc_horizon = repo.get_gc_horizon();
+
         // Garbage collect old files that are not needed for PITR anymore
         if gc_horizon > 0 {
-            repo.gc_iteration(None, gc_horizon, repo.get_pitr_interval(), false)?;
-        }
+            let gc_result = tokio::task::spawn_blocking(move || {
+                repo.gc_iteration(None, gc_horizon, repo.get_pitr_interval(), false)
+            })
+            .await;
 
-        // TODO Write it in more adequate way using
-        // condvar.wait_timeout() or something
-        let mut sleep_time = repo.get_gc_period().as_secs();
-        while sleep_time > 0 && tenant_mgr::get_tenant_state(tenantid) == Some(TenantState::Active)
-        {
-            sleep_time -= 1;
-            std::thread::sleep(Duration::from_secs(1));
+            match gc_result {
+                Ok(Ok(_gc_result)) => {
+                    // Gc success, do nothing
+                }
+                Ok(Err(e)) => {
+                    anyhow::bail!(e.context("Gc failed"));
+                }
+                Err(e) => {
+                    anyhow::bail!("Gc join error {}", e);
```

Comment:
NIT: The join error is now happens due to panic in the underlying task only, so we could get more specific.

---

IMO, we should not `bail!` in both `Err` branches, since previously both threads were specifically set to not to fail the system and they should try to finish their work, otherwise the resources will be depleted?